### PR TITLE
add support for podman create/run --domainname

### LINKED ISSUE #15102: Possibility to set domainname via REST Endpoint
**Is this a BUG REPORT or FEATURE REQUEST? (leave only one on its own line)**

/kind feature

**Description**

I'd like to implement possibility to [set domainname in podman-py project](https://github.com/containers/podman-py/blob/main/podman/domain/containers_create.py#L301) but according to the [podman REST API](https://docs.podman.io/en/latest/_static/api.html#tag/containers/operation/ContainerCreateLibpod) there is no such option.
Interestingly enough[ under compact (aka docker compatibility) endpoint there is such parameter](https://docs.podman.io/en/latest/_static/api.html#tag/containers-(compat)/operation/ContainerCreate), but podman-py (rightfully) does not use this endpoint to create container.

I'm also not seeing this option in cli (to be fair docker up until 1.40~ did not have this option either, but the call on the docker-py worked properly).

The inspect in podman shows the empty value, which makes me wonder whether this was forgotten or is there other option to fill the domainname?
```
[root ~]# podman inspect test --format '{{json .Config}}' | jq | head -n 3
{
  "Hostname": "test",
  "Domainname": "",

```

Is there possibility to add possibility to add domainname, to at least REST endpoint ?

### PR DESCRIPTION
**Pull request by cdoern**
domainname functions similarly to --hostname, it sets the value in /etc/hosts
to whatever you speficy in conjunction with --hostname. so if you pass
--hostname=foo --domainmame=baz.net, /etc/hosts will get the combined entry and /proc/sys/kernel/domainname will get baz.net via a sysctl.

resolves #15102

Signed-off-by: Charlie Doern <cdoern@redhat.com>

#### Does this PR introduce a user-facing change?

```release-note
introduce --domanname to podman create/run which allows users to set a domainname along with a hostname in their host files
```

### CODE REVIEW

File: cmd/podman/containers/create.go [hunk 1]
```diff
@@ -298,6 +298,10 @@ func CreateInit(c *cobra.Command, vals entities.ContainerCreateOptions, isInfra
 		vals.Entrypoint = &val
 	}
 
+	if !isInfra && c.Flag("domainname").Changed && !c.Flag("hostname").Changed {
+		return vals, fmt.Errorf("%w: cannot specify domainname without hostname", define.ErrInvalidArg)
+	}
+
```
Comment (Luap99) [inline cmd/podman/containers/create.go:304]: This does not match docker

File: pkg/specgen/generate/namespaces.go [hunk 2]
```diff
@@ -443,7 +447,11 @@ func specConfigureNamespaces(s *specgen.SpecGenerator, g *generate.Generator, rt
 		// the user or if we are creating a new UTS namespace.
 		// TODO: Should we be doing this for pod or container shared
 		// namespaces?
-		g.SetHostname(hostname)
+		if len(domainname) > 0 {
+			g.SetHostname(hostname + "." + domainname)
+		} else {
+			g.SetHostname(hostname)
+		}
```
Comment (Luap99) [inline pkg/specgen/generate/namespaces.go:454]: docker does not do this

**Review by Luap99 — CHANGES_REQUESTED** [review]
I took a quick look at docker and this doe snot match it.
The domain name is added in /etc/hosts and `setdomainname()`? Looks like they jusy use a sysctl for that: https://github.com/moby/moby/blob/master/daemon/oci_utils.go#L8 

Also you are ignoring the compat API value and it must be added to podman container inspect.

File: pkg/specgen/generate/namespaces.go (hunk 2 above)
Comment (cdoern) [reply pkg/specgen/generate/namespaces.go:454]: yeah you are right, I made a incorrect assumption about /etc/hostname. the  /etc/hosts behavior is correct though, I just need to add a default if domainname is specified and hostname is not

File: cmd/podman/containers/create.go (hunk 1 above)
Comment (cdoern) [reply cmd/podman/containers/create.go:304]: > I took a quick look at docker and this doe snot match it. The domain name is added in /etc/hosts and `setdomainname()`? Looks like they jusy use a sysctl for that: https://github.com/moby/moby/blob/master/daemon/oci_utils.go#L8
> 
> Also you are ignoring the compat API value and it must be added to podman container inspect.

that sysctl just takes care of the `setdomainname` aspect right? we still have to write to /etc/hosts as we do with other host vals?

Comment (cdoern) [reply cmd/podman/containers/create.go:304]: ahhhh we can't use the sysctl as rootless, you get: `Error: crun: open /proc/sys/kernel/domainname: Permission denied: OCI permission denied`

Comment (Luap99) [reply cmd/podman/containers/create.go:304]: Interesting setdomainname() seems to work:
```
unshare -r -u sh -c 'domainname test && domainname'
```

Comment (cdoern) [reply cmd/podman/containers/create.go:304]: @Luap99 see: https://github.com/moby/moby/pull/41840

we might need to make a crun change if we want to do it this way.

do you think we could just handle it manually rather than using the sysctl?

Comment (Luap99) [reply cmd/podman/containers/create.go:304]: sounds like something that would need to get added to the oci spec so that all oci runtimes have to do it

Manually calling setdomainname is possible but we would have to join the container uts namespace which is always difficult to do properly in go.

Comment (cdoern) [reply cmd/podman/containers/create.go:304]: https://github.com/containers/crun/blob/b66340c93cc8e8a5a8148ea840221b2cb663d896/tests/test_start.py#L186-L216 makes me think @giuseppe might have some reservations with this.. crun seems to think this should not be allowed

Comment (flouthoc) [reply cmd/podman/containers/create.go:304]: Not at sure but I think this can part of OCI spec, spec can have a special field and let runtime process this after setting up UTS namespace as edge-case or a small hack in runtime  could be just create/setup UTS namespace after `kernel.domainname` as being discussed here ( https://github.com/containers/crun/pull/983 ) but second one sounds like hack.

Comment (cdoern) [reply cmd/podman/containers/create.go:304]: so we'd have to modify the OCI spec? @flouthoc

**Comment by giuseppe** [comment]
I think the problem is with the `/proc/sys/kernel/domainname` file that is always owned by root in the initial user namespace, so writing to it always fails from a rootless user namespace, even if the value is namespaced.

I am not sure how the patch to crun can address it, have you verified it really writes to `/proc/sys/kernel/domainname` and that you can read the value back from `/proc/sys/kernel/domainname` within the container?

The correct fix IMO is to address it in the OCI specs and make domainname explicit as the hostname is.

The only possible workaround I see is to convert a write to `kernel.domainname` to a `setdomainname(2)`, something like:

```diff
diff --git a/src/libcrun/linux.c b/src/libcrun/linux.c
index 5a7531e..4cd39e2 100644
--- a/src/libcrun/linux.c
+++ b/src/libcrun/linux.c
@@ -3255,6 +3255,14 @@ libcrun_set_sysctl (libcrun_container_t *container, libcrun_error_t *err)
       if (UNLIKELY (ret < 0))
         return ret;
 
+      if (strcmp (name, "kernel/domainname") == 0)
+        {
+          ret = setdomainname (def->linux->sysctl->values[i], strlen (def->linux->sysctl->values[i]));
+          if (UNLIKELY (ret < 0))
+            return crun_make_error (err, errno, "setdomainname");
+          continue;
+        }
+
```

I am fine to such a workaround if it unblocks you, but I still suggest to 1) verify your patch really does what it is supposed to do 2) propose a fix for the OCI runtime specs.

Also no need to block with rootless if it is not currently supported.  Just skip the test.

**Comment by cdoern** [comment]
I agree with the OCI spec approach. I've been messing around with the crun hack all day and it doesn't seem to want to cooperate. I'll make a PR for the runtime spec on Monday.

**Comment by rhatdan** [comment]
Does this work with Docker?  Perhaps only because docker daemon runs as root and running docker in rootless mode this would fail also?

**Comment by mheon** [comment]
Do we want to add the flag as root-only for now, and enable rootless support once the OCI spec PR merges and support actually lands in crun/runc?

**Comment by cdoern** [comment]
> Do we want to add the flag as root-only for now, and enable rootless support once the OCI spec PR merges and support actually lands in crun/runc?

That sgtm, I haven't pushed the new version with the sysctl. I'll finish that up on Monday then take care of the spec PR

File: docs/source/markdown/options/domainname.md [hunk 3]
```diff
@@ -0,0 +1,5 @@
+#### **--domainname**=*name*
+
+Container domain name
+
+Sets the container domain name that is available inside the container. Can only be used with a private UTS namespace `--uts=private` (default).
```
Comment (rhatdan) [inline docs/source/markdown/options/domainname.md:5]: This should mention that currently this is only available in rootless mode.2

File: pkg/specgen/generate/namespaces.go [hunk 4]
```diff
@@ -450,8 +454,13 @@ func specConfigureNamespaces(s *specgen.SpecGenerator, g *generate.Generator, rt
 		// the user or if we are creating a new UTS namespace.
 		// TODO: Should we be doing this for pod or container shared
 		// namespaces?
+		// we do not want to add domainname to the hostname, we need to keep track of it for later
 		g.SetHostname(hostname)
 	}
+
+	if s.Domainname != "" {
+		g.AddLinuxSysctl("kernel.domainname", domainname)
+	}
```
Comment (giuseppe) [inline pkg/specgen/generate/namespaces.go:463]: your change for domainname got into the runtime-specs.  Why do not we use that (and also add support for it to crun)?

File: docs/source/markdown/podman-create.1.md.in [hunk 5]
```diff
@@ -91,6 +91,8 @@ each of stdin, stdout, and stderr.
 
 Block IO relative device weight.
 
+Block IO weight (relative device weight, format: `DEVICE_NAME:WEIGHT`).
+
```
Comment (giuseppe) [inline docs/source/markdown/podman-create.1.md.in:95]: it seems unrelated

File: pkg/specgen/generate/namespaces.go (hunk 4 above)
Comment (cdoern) [reply pkg/specgen/generate/namespaces.go:463]: ok, sounds good. There was initial discussion of getting this version in, but now that the spec PR has merged, I think we can wait on this.

**Comment by cdoern** [comment]
@giuseppe @Luap99 @rhatdan PTAL, once my PR in runtime-tools gets merged in I think this is the proper structure.

**Comment by cdoern** [comment]
@giuseppe can we publish a new crun release so these tests can work?

**Comment by giuseppe** [comment]
> @giuseppe can we publish a new crun release so these tests can work?

even if I do it today, it will still take some time it is available in Fedora (~1 week) and in Ubuntu.

I'd suggest we just skip the test for now and we can enable it once the OCI runtimes are ready

**Comment by rhatdan** [comment]
SGTM

**Comment by rhatdan** [comment]
@cdoern What is going on with this one. We have shipped a new crun?

**Comment by cdoern** [comment]
I'll fix the vendoring and skip the tests @rhatdan. I guess we can make some patches if this ends up not working once crun tests can work

**Comment by rhatdan** [comment]
@cdoern Needs a rebase.

File: pkg/specgen/generate/namespaces_linux.go [hunk 6]
```diff
@@ -98,8 +102,13 @@ func specConfigureNamespaces(s *specgen.SpecGenerator, g *generate.Generator, rt
 		// the user or if we are creating a new UTS namespace.
 		// TODO: Should we be doing this for pod or container shared
 		// namespaces?
+		// we do not want to add domainname to the hostname, we need to keep track of it for later
 		g.SetHostname(hostname)
 	}
+
+	if s.Domainname != "" || s.UtsNS.NSMode != specgen.Host {
```
Comment (rhatdan) [inline pkg/specgen/generate/namespaces_linux.go:109]: Should this be an &&?

Otherwise you could be setting Domain name to ""?

Comment (cdoern) [reply pkg/specgen/generate/namespaces_linux.go:109]: right, I will change that

**Comment by cdoern** [comment]
@giuseppe @rhatdan  PTAL

File: docs/source/markdown/options/domainname.md [hunk 7]
```diff
@@ -0,0 +1,9 @@
+####> This option file is used in:
+####>   podman create, run
+####> If file is edited, make sure the changes
+####> are applicable to all of those.
+#### **--domainname**=*name*
+
+Container domain name
+
+Sets the container domain name that is available inside the container. Can only be used with a private UTS namespace `--uts=private` (default). This option is only curently only available in rootful mode.
```
Comment (giuseppe) [inline docs/source/markdown/options/domainname.md:9]: typo in `curently`

Why is it only available for root?  We run the tests we have added for crun also with unprivileged users

File: test/e2e/run_dns_test.go [hunk 8]
```diff
@@ -99,6 +99,19 @@ var _ = Describe("Podman run dns", func() {
 		Expect(session.OutputToString()).To(ContainSubstring("foobar"))
 	})
 
+	It("podman run domainname sets /etc/hosts and /proc/sys/kernel/domainname", func() {
+		Skip("Waiting for crun release > 1.6")
```
Comment (giuseppe) [inline test/e2e/run_dns_test.go:103]: crun 1.6 was released some time ago.  Where is the test failing?

Comment (cdoern) [reply test/e2e/run_dns_test.go:103]: ah, let me un-comment that and see if this all works...

File: docs/source/markdown/options/domainname.md (hunk 7 above)
Comment (cdoern) [reply docs/source/markdown/options/domainname.md:9]: leftover from the first implementation, ill change

**Comment by cdoern** [comment]
@giuseppe it seems /proc/sys/kernel/domainname does not get set but /etc/hosts does. Does this mean the domainname is working or not? not sure if the /proc file is meant to be set

**Comment by giuseppe** [comment]
> @giuseppe it seems /proc/sys/kernel/domainname does not get set but /etc/hosts does. Does this mean the domainname is working or not? not sure if the /proc file is meant to be set

it sounds like working from Podman but it doesn't look like working from the OCI runtime.  To what test are you referring?

**Comment by cdoern** [comment]
> > @giuseppe it seems /proc/sys/kernel/domainname does not get set but /etc/hosts does. Does this mean the domainname is working or not? not sure if the /proc file is meant to be set
> 
> it sounds like working from Podman but it doesn't look like working from the OCI runtime. To what test are you referring?

@giuseppe I was checking /proc/sys/kernel/domainname in a test but I am not sure if that file is meant to contain the --domainname. running commands like `hostname -A` work appropriately though... so i am not sure if crun is doing it right or not.

**Comment by rhatdan** [comment]
```
# domainname dan
sh-5.2# domainname
dan
sh-5.2# cat /proc/sys/kernel/domainname 
dan
```

If I run a container I will also see domainname dan within it, which seems wrong, but I can change it


```
sh-5.2# podman run -ti --privileged fedora sh
# dnf -y install /usr/bin/domainname
...
Installed:
  hostname-3.23-7.fc37.x86_64                                                                                                                                                        

Complete!
# domainname
dan
# domainname walsh
# domainname
walsh
# exit
sh-5.2# domainname
dan
```

**Comment by rhatdan** [comment]
I would expect the /proc/sys/kernel/domainname  to show the containers domainname, if this is set.

**Comment by cdoern** [comment]
> I would expect the /proc/sys/kernel/domainname to show the containers domainname, if this is set.

ok, then I think this is an issue with crun. Since /etc/hosts seems to be updated and the spec does indeed have the proper domainname entity set, something must be wrong..

though, in crun all of the domainname tests pass. @giuseppe WDYT?

**Comment by giuseppe** [comment]
is the correct crun installed?

If I pull locally your changes (testing commit c91508b9c18a3f193c7ab28beaaad85573747e3d), I see it works:

```
$ sudo bin/podman run --rm --domainname as-root fedora cat /proc/sys/kernel/domainname
as-root
$ bin/podman run --rm --domainname and-also-rootless fedora cat /proc/sys/kernel/domainname
and-also-rootless
```

Instead of installing an additional package, let's change the test to print directly `/proc/sys/kernel/domainname`

**Comment by cdoern** [comment]
@giuseppe could ci have an old crun version?

**Comment by giuseppe** [comment]
yes I think that is the issue.  It is using crun-1.6-2.fc36-x86_64 while the support was added in crun 1.7.

Let's not block on the CI, just skip the test for now

**Comment by vrothberg** [comment]
@cdoern I think rebasing and unskipping the test is possible now.   Do you have time to tackle it?

**Comment by vrothberg** [comment]
Friendly ping.  @cdoern I think we are good to give it another try.
